File: php/features.php
<?php
declare(strict_types=1);

// WebsiteCarbon SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class WebsiteCarbonFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new WebsiteCarbonBaseFeature();
            case "ratelimit":
                return new WebsiteCarbonRatelimitFeature();
            case "retry":
                return new WebsiteCarbonRetryFeature();
            case "test":
                return new WebsiteCarbonTestFeature();
            case "timeout":
                return new WebsiteCarbonTimeoutFeature();
            default:
                return new WebsiteCarbonBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/websitecarbon_sdk/features.py
# WebsiteCarbon SDK feature factory

from websitecarbon_sdk.feature.base_feature import WebsiteCarbonBaseFeature
from websitecarbon_sdk.feature.test_feature import WebsiteCarbonTestFeature


def _make_feature(name):
    features = {
        "base": lambda: WebsiteCarbonBaseFeature(),
        "test": lambda: WebsiteCarbonTestFeature(),
    }
    factory = features.get(name)
    if factory is not None:
        return factory()
    return features["base"]()
